xrchat/xrchat-server
# Seats

**What is the purpose of this PR?**
Implement seats for a subscription.

**What does issue #s are solved by this PR?**
No specific issues

**What unit tests have been put in place?**
None

**What files or features are including in this PR that are not related to the main feature?**
None

**How does our QA team test this PR?**
This is most easily tested with the client branch 'seats'.

Login with a user, then sign up for a plan. It's best to sign up for something above 'Journey', as that only has one seat.

Subscriptions now have four new column: totalSeats, unusedSeats, pendingSeats, and filledSeats. Respectively, they are: the number of seats the subscription is entitled to; the number of seats that are neither pending nor filled; the number of invitations that have been sent out but not confirmed; and the number of invitations that have been confirmed (plus a seat for the owner of the subscription, which will automatically be created and cannot be removed).

Go to the Profile modal, and then go to the Subscription tab. There should now be a button 'Manage Subscription Seats'. Click that to go to /subscription/seats, a new page.

If the user has a subscription, they should see their account occupying a seat. This seat cannot be removed.

Enter an email address into the text field and either press enter or click the button next to it. This will do a magicLink login for that address - if it's already in the system, no new user/identityProvider will be created, but if it's not, a new user/email identityProvider will be made. The email that gets sent out has an additional query parameter for subscriptionId.

At this point, the email address should show up above the self user. You can press the 'Cancel Invitation' button to remove the pending seat. If you follow the login link, it should not do anything to the seat status of that subscription.

If you follow the login link when there is a pending seat for it, the user should be logged in and the pending seat should be converted to a filled seat. If you follow the link again for a user of a filled seat, nothing more seat-wise should happen.

As the subscription holder, you can revoke filled seats by pressing the button next to those users. If you've remained on the seats page the whole time and filled the seat through, say, an incognito window, you'll have to refresh the page to see the button switch from 'cancel invitation' to 'revoke seat', as the page currently isn't subscribing to any changes from the server.

NOTE:
Patching seats is a little different than expected. Since going from pending->filled is going to take place when the subscription owner isn't present, the ID for this should be the userId of the seat and the data that's passed in is just { subscriptionId: <subscriptionId> }. The server checks to see if that user has a pending seat on that subscription and converts it to filled if that's the case.
